presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java

```diff
@@ -283,6 +284,11 @@ public SystemSessionProperties(
                         featuresConfig.isLegacyOrderBy(),
                         false),
                 booleanSessionProperty(
+                        SUPPORT_NULL_ON_LHS_OF_SEMI_JOIN,
```

Comment:
> I also added a TODO in EffectivePredicateExtractor where it is incorrect (projections and nulls)

AFAIR predicates extracted from `EffectivePredicateExtractor` are only used for equality inference. Equality inference determines if it is possible to replace one symbol with some equivalence. In this context `EffectivePredicateExtractor` works correctly. Predicate `a=null` extracted from projection means it is is possible to use `null` in all places where `a` would be used (e.g: `f(a)` can be rewritten as `f(null)`), because those values will always be the same in row context. You can think of more complex examples, e.g: projection `a=b+3`. In this case it also works, when `b` is `null`, `a` is `null` too. Because of that I think `EffectivePredicateExtractor` name is misleading. I think it should actually be called `EffectiveEquvalenceExtractor`.

If you fixed the projection as you mentioned, then it would basically break equality inference for predicated derived from projections. This is because you would have to append `or x is null` whenever you are not sure that the symbol value won't be `null`, (e.g: predicate derived from projection `a=b+3` would be `a=b+3 or (a is null)`). Since equality inference doesn't process alternatives it wouldn't use those `ORed` predicates.
Probably we need two mechanisms:
1. `EffectiveEquvalenceExtractor` which is basically current `EffectivePredicateExtractor` (probably more explicit)
2. Some other mechanism to determine nullability of a symbol. Maybe nullability should be a kind of trait. I've seen that calcite has nullability as type parameter. There is also basic nullability analyzer for expressions in Presto (`NullabilityAnalyzer`).

> Maybe fixing EffectivePredicateExtractor is what it takes to re-enable certain pushdowns.

This won't solve all issues, e.g:
`SELECT * FROM ... WHERE (x = 3 OR x is null) AND (x IN (...))`
It doesn't matter that 'x' can be null. It doesn't matter since `IN` is used in filter.

On the other hand:
`SELECT (x IN (...)) FROM ... WHERE (x = 3)`
without nullability analysis we won't push predicate down to semi join.

So we probably need both nullability analysis and null invariance analysis.